core/domain/learner_progress_services.py

```diff
@@ -0,0 +1,268 @@
+# coding: utf-8
+#
+# Copyright 2014 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Services for tracking the progress of the learner."""
+
+from core.domain import subscription_services
+from core.platform import models
+import utils
+
+(user_models,) = models.Registry.import_models([models.NAMES.user])
+
+
+def add_exp_id_to_completed_list(user_id, exp_id):
+    """Adds the exploration id to the completed list of the user unless the
+    exploration has already been completed or has been created by the user.
+    It is also removed from the incomplete list (if present).
+
+    Args:
+        user_id: str. The id of the user who has completed the exploration.
+        exp_id: str. The id of the completed exploration.
+    """
+    activities_completed_model = (
+        user_models.ActivitiesCompletedByLearnerModel.get(
+            user_id, strict=False))
+    if not activities_completed_model:
+        activities_completed_model = (
+            user_models.ActivitiesCompletedByLearnerModel(id=user_id))
+
+    exploration_created_ids = (
+        subscription_services.get_exploration_ids_subscribed_to(user_id))
+
+    if (exp_id not in exploration_created_ids and
+            exp_id not in activities_completed_model.completed_exploration_ids):
```

Comment:
I don't quite understand this. I looked at some examples in the codebase which used domain layer object and it seems like the domain layer objects are only used to "get" values. For updating values, we have directly fetched the model and updated it. For example in user_services.py in `get_email_preferences_for_exploration` we have used to the user domain object. But in `set_email_preferences_for_exploration` we have fetched the storage layer instance and updated it. Similarly in collection_services in `_save_collection` we have fetched the storage instance and updated it (lines 695 to 706). So I'm a bit confused on the last step `save the domain object (after converting it to storage model` as it seems like we still depend on the storage layer instances.